--- kmp.h.orig	2025-05-20 19:15:00.633881763 +0300
+++ kmp.h	2025-05-20 19:47:47.770195902 +0300
@@ -11,16 +11,20 @@
 //
 //===----------------------------------------------------------------------===//

 #ifndef KMP_H
 #define KMP_H

 #include "kmp_config.h"

+#if defined(__has_feature) && __has_feature(memory_sanitizer)
+#include <sanitizer/msan_interface.h>
+#endif
+
 /* #define BUILD_PARALLEL_ORDERED 1 */

 /* This fix replaces gettimeofday with clock_gettime for better scalability on
    the Altix.  Requires user code to be linked with -lrt. */
 //#define FIX_SGI_CLOCK

 /* Defines for OpenMP 3.0 tasking and auto scheduling */

@@ -1414,21 +1418,27 @@ extern void __kmp_query_cpuid(kmp_cpuinf
 // zero in most cases
 static inline void __kmp_x86_cpuid(int leaf, int subleaf, struct kmp_cpuid *p) {
   __asm__ __volatile__("cpuid"
                        : "=a"(p->eax), "=b"(p->ebx), "=c"(p->ecx), "=d"(p->edx)
                        : "a"(leaf), "c"(subleaf));
 }
 // Load p into FPU control word
 static inline void __kmp_load_x87_fpu_control_word(const kmp_int16 *p) {
+#if defined(__has_feature) && __has_feature(memory_sanitizer)
+  __msan_check_mem_is_initialized(p, sizeof(*p));
+#endif
   __asm__ __volatile__("fldcw %0" : : "m"(*p));
 }
 // Store FPU control word into p
 static inline void __kmp_store_x87_fpu_control_word(kmp_int16 *p) {
   __asm__ __volatile__("fstcw %0" : "=m"(*p));
+#if defined(__has_feature) && __has_feature(memory_sanitizer)
+  __msan_unpoison(p, sizeof(*p));
+#endif
 }
 static inline void __kmp_clear_x87_fpu_status_word() {
 #if KMP_MIC
   // 32-bit protected mode x87 FPU state
   struct x87_fpu_state {
     unsigned cw;
     unsigned sw;
     unsigned tw;
